text/0000-repr-type-aliases.md

```diff
@@ -0,0 +1,84 @@
+
+You can use any type alias in the `repr` attribute, but it *must* be an alias to an accepted primitive type like `u8` or `i32`, and cannot be a pointer, reference, struct, etc.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The `repr` attribute now accepts a `type = ...` argument to indicate a resolved path instead of a well-known primitive type. If those the path resolves to a type alias to a valid primitive type which can be used in the `repr` attribute, that will be used as the actual discriminant representation.
+
+An additional, automatically-applicable lint should be added that warns a user if they use `type = ...` for a well-known primitive type, since adding `type = ` instead of using the type directly introduces the possibility of shadowing. (For example, `#[repr(type = u32)]` becomes `#[repr(u32)]`.)
+
+Similarly, an automatically-applicable lint should be added that warns a user if a `repr` argument references in an-scope type alias without the `type = ` prefix. (For example, `#[repr(MyType)]` becomes `#[repr(type = MyType)]`.)
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The requirement for `type =` is unfortunate, but it feels like the best way to ensure that adding new representations isn't a breaking change going forward. Even if we were to decide it weren't a "breaking change," it would still break things anyway, being de-facto breaking.
+
+And, of course, this complicates the compiler. But that's about it.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+We could always not do this.
+
+But more realistically, here are some alternative designs that were rejected.
+
+## `self::`
+
+We could, instead of using `type =`, require that all types contain a double-colon to indicate they're a path, effectively preventing collisions with arguments that aren't paths. This would require using `self::` for types that are imported in the local scope, and was actually the first proposal of this RFC, but wasn't very well-received.
+
+## Shadowing attributes
+
+Until a future edition, the current set of valid representations could be solidified as taking precedence over any shadowed identifiers. For example, if someone defines `type transparent = u32`, then `repr(transparent)` still means `repr(transparent)` and not `repr(u32)`.
+
+In future editions, we could either:
+
+* Let type aliases shadow all valid representations. This isn't ideal since there is no way to override the shadowing besides nesting your code in a new module and then re-exporting it outside that module, which is very messy.
+* Expand the list of unshadowable representations every edition where necessary.
+
+## Capital letters
+
+You could require that the types start with capital letters- oh, right, `repr(C)` is a thing.
```

Comment:
I'm confused what this comment means; could you elaborate?

Reply:
We cannot require that types start with capital letters, because types with lower-case first letters can and do exist. Among these are the primitive integer types (which presumably would be special-case allowed), but there are absolutely cases of lower-case types in the ecosystem (e.g., [`f16`](https://docs.rs/half/2.4.1/half/struct.f16.html)). It would intensely odd for such types to be responded to with a parse error.

Reply:
Sorry, I should have clarified: what I mean is that types with capital letters in them would be specifically interpreted as type aliases, and that the existing ones would still be allowed. It's sort of a half-hearted attempt to point out that this would be a terrible idea, but I should probably just remove it.